docs/solidity-by-example.rst

```diff
@@ -645,4 +645,509 @@ Safe Remote Purchase
+at the time of contract deployment. Once that time is reached, the Alice can call
+*claimTimeout* to recover their funds. You can see the *claimTimeout* function in the
+full contract.
+
+After this function is called, the Bob can no longer receive any Ether,
+so it is important that the Bob close the channel before the expiration is reached.
+
+
+The full contract
+-----------------
+
+::
+
+    pragma solidity ^0.4.24;
+
+    contract SimplePaymentChannel {
+        address public sender;      // The account sending payments.
+        address public recipient;   // The account receiving the payments.
+        uint256 public expiration;  // Timeout in case the recipient never closes.
+
+        constructor (address _recipient, uint256 duration)
+            public
+            payable
+        {
+            sender = msg.sender;
+            recipient = _recipient;
+            expiration = now + duration;
+        }
+
+        function isValidSignature(uint256 amount, bytes signature)
+        internal
+        view
+        returns (bool)
+        {
+            bytes32 message = prefixed(keccak256(abi.encodePacked(this, amount)));
+
+            // check that the signature is from the payment sender
+            return recoverSigner(message, signature) == sender;
+        }
+
+        /// the recipient can close the channel at any time by presenting a
+        /// signed amount from the sender. the recipient will be sent that amount,
+        /// and the remainder will go back to the sender
+        function close(uint256 amount, bytes signature) public {
+            require(msg.sender == recipient);
+            require(isValidSignature(amount, signature));
+
+            recipient.transfer(amount);
+            selfdestruct(sender);
+        }
+
+        /// the sender can extend the expiration at any time
+        function extend(uint256 newExpiration) public {
+            require(msg.sender == sender);
+            require(newExpiration > expiration);
+
+            expiration = newExpiration;
+        }
+
+        /// if the timeout is reached without the recipient closing the channel,
+        /// then the Ether is realeased back to the sender.
+        function clainTimeout() public {
+            require(now >= expiration);
+            selfdestruct(sender);
+        }
+
+        /// from here to the end of this contract, all the functions we already wrote, in
+        /// the 'creating and verifying signatures' chapter, so if you already know what them
+        /// does, you can skip it.
+
+        /// the same functions we wrote in the 'creating and verifying signatures' chapter,
+        /// you can go there to find the full explanations
+
+        function splitSignature(bytes sig)
+            internal
+            pure
+            returns (uint8, bytes32, bytes32)
+        {
+            require(sig.length == 65);
+
+            bytes32 r;
+            bytes32 s;
+            uint8 v;
+
+            assembly {
```

Comment:
I wouldn't mind an explanation here this code is from https://gist.github.com/axic/5b33912c6f61ae6fd96d6c4a47afde6d, but in a real implementation a more tested library should be used, such as openzepplin's fork of this code: https://github.com/OpenZeppelin/openzeppelin-solidity/blob/master/contracts/ECRecovery.sol

Reply:
Sorry, I will put the explanation and original/fork links.

Reply:
I don't know if I can use the library inside this docs section, so I put a note below the contract.